report.php

```diff
@@ -7,27 +7,57 @@
 
 use ITBugTracking\Factories\DatabaseConnector;
 use ITBugTracking\Hydrators\IssueHydrator;
+use ITBugTracking\Services\ValidationService;
 
 $db = DatabaseConnector::connect();
 
 $json = file_get_contents('php://input');
 
 $data = json_decode($json, true);
 
-$newIssue = IssueHydrator::createIssue($db, $data);
+$requiredExist = ValidationService::checkRequiredDataExists($data);
+$validTitle = ValidationService::limitTitleCharLengthTo255($data['title']);
+$validReporter = ValidationService::limitReporterCharLengthTo255($data['name']);
+$data['description'] = ValidationService::descriptionLimitCharLength($data['description']);
+//$validSeverity = ValidationService::checkSeverityExists($data['severity']);
+$data['severity'] = ValidationService::checkSeverityIsInt($data['severity']);
+$checkDepartmentIsInt = ValidationService::checkDepartmentIsInt($data['department']);
+echo '<pre>';
+var_dump($checkDepartmentIsInt);
+$passedValidation = $requiredExist && $validTitle && $validReporter && $data['severity'] && $checkDepartmentIsInt;
+
+if (!$passedValidation) {
+    http_response_code(400);
+    echo json_encode(["message" => "Validation failed"]);
+    return;
+} else {
```

Comment:
```suggestion
// first tip, space it out - one newline between unrelated things helps you keep track of what you're doing

// track the result of true/false validations in variables
$requiredExist = ValidationService::checkRequiredDataExists($data);

// if we don't have the required fields then we can't really do any of the other validations, so we would need an early out
//if (!$requiredExist) {
//    http_response_code(400);
//    echo json_encode(["message" => "Invalid issue data"]);  // make sure to use the proper message fromt he API spec
//    return;  // could also use exit() to ensure an early exit
//}

// if it is a sanitisation that changes the data in some way, assign it to the data
// the product owner may not want you to actively change the user's input data - that's a risky path
// they may prefer you just reject data that won't fit into the database
// you need to ask them if it isn't explicit in the story
$data['title'] = ValidationService::limitStringLength($data['title'], 255);
// remember html special chars on strings to reduce the ability of our users to input naughty strings
$data['title'] = htmlspecialchars($data['title']);

$data['name'] = ValidationService::limitStringLength($data['name'], 255);
$data['name'] = htmlspecialchars($data['name']);

$data['description'] = ValidationService::limitStringLength($data['description'], 10000);
$data['description'] = htmlspecialchars($data['description']);
// if you want to streamline this repeated code, then that is a candidate for a re-usable function
// you need to write the code first before you can figure out what you do again and again
// but validations just are kind of repetitive - it's really not a big deal to follow the same procedure a few times to validate string fields

$severityIsNumeric = is_numeric($data['severity']);
if ($severityIsNumeric) {
    $data['severity'] = (int)$data['severity'];  // this is how we can convert anything that passes an is_numeric check to int in PHP
    $severityExists = ValidationService::checkSeverityExists($db, $data['severity']);
} else {
    $severityExists = false;
}

$departmentIsNumeric = is_numeric($data['department']);
if ($departmentIsNumeric) {
    $data['department'] = (int)$data['department'];
    $departmentExists = ValidationService::checkDepartmentExists($db, $data['department']);
} else {
    $departmentExists = false;
}

$passedValidation = 
    $data['title']
    && $data['name'] 
    // && $data['description']  // description is a nullable field, so maybe we don't need to check if it exists!
    && $severityExists 
    && $departmentExists;

if (!$passedValidation) {
    http_response_code(400);
    echo json_encode(["message" => "Invalid issue data"]);
    return;  // could also use exit()
} else {
```

Reply:
I have checked the frontend and the max lengths for:
name => 255
title => 100
description => 65535
Since this is encoded in the frontend, I think it's safe to reject input that exceeds these limits.

Reply:
done